### PR TITLE
feature: gather VM metrics and forward to AppSignal

### PR DESCRIPTION
**Pull request by InoMurko**
An example dashboard for the current metrics.
```
-
  title: 'VM dashboard'
  graphs:
    -
      title: 'VM atom over time'
      kind: gauge
      format: number
      draw_null_as_zero: true
      fields:
        - erlang_system_info_atom_count
      tags:
        node: '*'
    -
      title: 'ETS count'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_ets_count
      tags:
        node: '*'
    -
      title: 'IO Input KB'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_io_input_kb
      tags:
        node: '*'        
    -
      title: 'IO Output KB'
      kind: gauge
      format: number
      draw_null_as_zero: true
      fields:
        - erlang_io_output_kb
      tags:
        node: '*'        
    -
      title: 'Port count'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_system_info_port_count
      tags:
        node: '*'        
    -
      title: 'Process count'
      kind: gauge
      format: number
      draw_null_as_zero: true
      fields:
        - erlang_system_info_process_count
      tags:
        node: '*'        
    -
      title: 'Number of schedulers'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_system_info_schedulers
      tags:
        node: '*'        
    -
      title: 'Erlang Total Run Queue Lengths'
      kind: gauge
      format: number
      draw_null_as_zero: true
      fields:
        - erlang_total_run_queue_lengths
      tags:
        node: '*'        
    -
      title: Uptime
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_uptime
      tags:
        node: '*'        
    -
      title: 'Atom memory'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_memory_atom
      tags:
        node: '*'        
    -
      title: 'Atom memory used'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_memory_atom_used
      tags:
        node: '*'        
    -
      title: 'Binary memory'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_memory_binary
      tags:
        node: '*'        
    -
      title: 'ETS memory'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_memory_ets
      tags:
        node: '*'       
    -
      title: 'Memory processes'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_memory_processes
      tags:
        node: '*'        
    -
      title: 'Memory total'
      kind: gauge
      draw_null_as_zero: false
      fields:
        - erlang_memory_total
      tags:
        node: '*'        

```

### CODE REVIEW

File: apps/omg_status/lib/status.ex [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  end
+
+  @spec vm_metrics :: maybe_improper_list(atom(), fun()) | []
+  defp vm_metrics, do: do_vm_metrics(is_enabled?())
+
+  defp do_vm_metrics(false), do: []
+
+  defp do_vm_metrics(true) do
+    memory =
+      for n <- ~w(total processes ets binary atom atom_used)a,
```
Comment (paulperegud) [inline apps/omg_status/lib/status.ex]: TIL

File: apps/omg_status/lib/status.ex [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          do: {String.to_atom("erlang_memory_#{n}"), fn -> :erlang.memory(n) end}
+
+    system_info =
+      for n <- ~w(schedulers atom_count process_count port_count)a,
```
Comment (paulperegud) [inline apps/omg_status/lib/status.ex]: n -> type ?

**Comment by InoMurko** [comment]
Hey @jbunce please take a look. There's both config configuration available and ENV variables. You can adjust the interval rate of each metric (do you need to be able to adjust it in runtime?). 
config var is an atom
sys var is an upper case string
```
:erlang_memory_total_interval
"ERLANG_MEMORY_TOTAL_INTERVAL"
:erlang_memory_processes_interval
"ERLANG_MEMORY_PROCESSES_INTERVAL"
:erlang_memory_ets_interval
"ERLANG_MEMORY_ETS_INTERVAL"
:erlang_memory_binary_interval
"ERLANG_MEMORY_BINARY_INTERVAL"
:erlang_memory_atom_interval
"ERLANG_MEMORY_ATOM_INTERVAL"
:erlang_memory_atom_used_interval
"ERLANG_MEMORY_ATOM_USED_INTERVAL"
:erlang_system_info_schedulers_interval
"ERLANG_SYSTEM_INFO_SCHEDULERS_INTERVAL"
:erlang_system_info_atom_count_interval
"ERLANG_SYSTEM_INFO_ATOM_COUNT_INTERVAL"
:erlang_system_info_process_count_interval
"ERLANG_SYSTEM_INFO_PROCESS_COUNT_INTERVAL"
:erlang_system_info_port_count_interval
"ERLANG_SYSTEM_INFO_PORT_COUNT_INTERVAL"
:erlang_uptime_interval
"ERLANG_UPTIME_INTERVAL"
:erlang_io_input_kb_interval
"ERLANG_IO_INPUT_KB_INTERVAL"
:erlang_io_output_kb_interval
"ERLANG_IO_OUTPUT_KB_INTERVAL"
:erlang_total_run_queue_lengths_interval
"ERLANG_TOTAL_RUN_QUEUE_LENGTHS_INTERVAL"
:erlang_ets_count_interval
"ERLANG_ETS_COUNT_INTERVAL"
```

**Comment by jbunce** [comment]
This is great. One nit: "erlang_total_run_queue_lengths" title of the dashboard should be  "Erlang Total Run Queue Lengths" in AppSignal.

> You can adjust the interval rate of each metric
Is the default 5000 milliseconds for sampling each metric? That sounds reasonable as a first pass. We can review based on what we discover operationally :)

File: apps/omg_status/lib/status.ex [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# Copyright 2018 OmiseGO Pte Ltd
```
Comment (jbunce) [inline apps/omg_status/lib/status.ex]: Nit: 2019

File: apps/omg_status/lib/status/metric/recorder.ex [hunk 4]
```diff
@@ -0,0 +1,77 @@
+# Copyright 2018 OmiseGO Pte Ltd
```
Comment (jbunce) [inline apps/omg_status/lib/status/metric/recorder.ex]: Same

File: apps/omg_status/test/test_helper.exs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2018 OmiseGO Pte Ltd
```
Comment (jbunce) [inline apps/omg_status/test/test_helper.exs]: 2019

**Comment by InoMurko** [comment]
> This is great. One nit: "erlang_total_run_queue_lengths" title of the dashboard should be "Erlang Total Run Queue Lengths" in AppSignal.
> 
> > You can adjust the interval rate of each metric
> 
> Is the default 5000 milliseconds for sampling each metric? That sounds reasonable as a first pass. We can review based on what we discover operationally :)

Yeah, each is 5seconds at the moment.

**Pull request merged**